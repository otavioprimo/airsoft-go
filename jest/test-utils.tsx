import React from 'react';
import {render} from '@testing-library/react-native';
import {ThemeProvider} from 'styled-components/native';
import {createStore, Reducer} from '@reduxjs/toolkit';
import {Provider} from 'react-redux';

import * as Theme from '../src/theme';

interface ReduxProviders {
  state: any;
  reducer: Reducer<any, any>;
}

interface RenderProviders {
  children: React.ReactElement;
  options?: ReduxProviders;
}

const ReduxProvider = ({children, options}: any) => {
  if (!options) {
    return <>{children}</>;
  }

  const store = createStore(options.reducer, options.state);
  return <Provider store={store}>{children}</Provider>;
};

const AllTheProviders: React.FC<RenderProviders> = ({children, options}) => {
  return (
    <ThemeProvider theme={Theme.DarkTheme}>
      <ReduxProvider options={options}>{children}</ReduxProvider>
    </ThemeProvider>
  );
};

const customRender = (ui: React.ReactElement, options?: ReduxProviders) => {
  return render(<AllTheProviders options={options}>{ui}</AllTheProviders>);
};

// re-export everything
export * from '@testing-library/react-native';

// override render method
export {customRender as render};
