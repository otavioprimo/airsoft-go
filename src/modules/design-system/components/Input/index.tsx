import React, {useState, useEffect} from 'react';
import {
  ContainerInput,
  ContainerIcon,
  Icon,
  InputText,
  ContainerText,
  Label,
} from './style';
import {TextInputProps} from 'react-native';
import {
  useSharedValue,
  useAnimatedStyle,
  withTiming,
  Easing,
} from 'react-native-reanimated';

interface InputProps extends TextInputProps {
  leftIcon?: any;
  label: string;
  ref?: any;
}

const TEST_ID_RIGHT_ICON = 'right_icon';
const TEST_ID_INPUT = 'input';

const Input: React.FC<InputProps> = React.forwardRef(
  (props: InputProps, ref: any) => {
    const LABEL_TRANSLATE_Y_FOCUSED = -10;
    const LABEL_TRANSLATE_Y_FOCUSED_OUT = 10;

    const LABEL_TRANSLATE_X_FOCUSED = props.leftIcon ? -30 : 0;
    const LABEL_TRANSLATE_X_FOCUSED_OUT = 0;

    const LABEl_FONT_SIZE_FOCUSED = 10;
    const LABEl_FONT_SIZE_FOCUSED_OUT = 14;

    const [focused, setFocused] = useState(false);
    const [typed, setTyped] = useState(false);

    const labelTranslateX = useSharedValue(LABEL_TRANSLATE_X_FOCUSED_OUT);
    const labelTranslateY = useSharedValue(LABEL_TRANSLATE_Y_FOCUSED_OUT);
    const labelFontSize = useSharedValue(LABEl_FONT_SIZE_FOCUSED_OUT);

    const animatedTranslateStyles = useAnimatedStyle(() => {
      return {
        transform: [
          {
            translateY: withTiming(labelTranslateY.value, {
              duration: 150,
              easing: Easing.out(Easing.linear),
            }),
          },
          {
            translateX: withTiming(labelTranslateX.value, {
              duration: 150,
              easing: Easing.out(Easing.linear),
            }),
          },
        ],
      };
    });

    const animatedFontStyle = useAnimatedStyle(() => {
      return {
        fontSize: withTiming(labelFontSize.value, {
          duration: 150,
          easing: Easing.out(Easing.linear),
        }),
      };
    });

    useEffect(() => {
      const startAnimation = typed || focused;
      labelTranslateX.value = startAnimation
        ? LABEL_TRANSLATE_X_FOCUSED
        : LABEL_TRANSLATE_X_FOCUSED_OUT;

      labelTranslateY.value = startAnimation
        ? LABEL_TRANSLATE_Y_FOCUSED
        : LABEL_TRANSLATE_Y_FOCUSED_OUT;

      labelFontSize.value = startAnimation
        ? LABEl_FONT_SIZE_FOCUSED
        : LABEl_FONT_SIZE_FOCUSED_OUT;
    }, [focused, typed]);

    return (
      <ContainerInput>
        {props.leftIcon && (
          <ContainerIcon testID={TEST_ID_RIGHT_ICON}>
            <Icon icon={props.leftIcon} />
          </ContainerIcon>
        )}
        <ContainerText>
          {props.label && (
            <Label style={[animatedTranslateStyles, animatedFontStyle]}>
              {props.label}
            </Label>
          )}

          <InputText
            testID={TEST_ID_INPUT}
            {...props}
            onFocus={() => setFocused(true)}
            onBlur={() => setFocused(false)}
            onChangeText={text => {
              setTyped(Boolean(text));
              if (props.onChangeText) {
                props.onChangeText(text);
              }
            }}
            underlineColorAndroid="transparent"
            ref={ref}
          />
        </ContainerText>
      </ContainerInput>
    );
  },
);

export default Input;
