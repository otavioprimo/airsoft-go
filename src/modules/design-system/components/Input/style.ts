import Animated from 'react-native-reanimated';
import styled from 'styled-components/native';
import {FontAwesomeIcon} from '@fortawesome/react-native-fontawesome';

export const ContainerInput = styled.View`
  flex-direction: row;
  padding-bottom: 16px;
`;

export const InputText = styled.TextInput.attrs(({theme}) => ({
  placeholderTextColor: theme.secondary,
}))`
  color: ${({theme}) => theme.textPrimary};
  font-size: 16px;
  border-bottom-color: ${({theme}) => theme.secondary};
  border-bottom-width: 1px;
  flex: 1;
`;

export const ContainerIcon = styled.View`
  padding: 8px;
  justify-content: center;
  align-items: center;
  border-bottom-width: 1px;
  border-bottom-color: ${({theme}) => theme.secondary};
`;

export const Icon = styled(FontAwesomeIcon).attrs(({theme}) => ({
  size: 19,
  color: theme.textPrimary,
}))``;

export const ContainerText = styled.View`
  height: 38px;
  flex: 1;
`;

export const Label = styled(Animated.Text)`
  color: ${({theme}) => theme.textPrimary};
  position: absolute;
`;
