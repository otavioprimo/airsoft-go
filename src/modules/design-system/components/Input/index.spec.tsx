import React from 'react';
import {render, fireEvent, act} from 'test/test-utils';
import {faEnvelope} from '@fortawesome/free-regular-svg-icons';

import Input, {TEST_ID_INPUT} from './index';

beforeAll(() => {
  jest.useFakeTimers();
});

describe('Design System - Input', () => {
  it('Should match snapshot', () => {
    const {toJSON} = render(<Input label="Email" />);
    expect(toJSON()).toMatchSnapshot('Input_default');
  });

  it('Should match snapshot with leftIcon', () => {
    const {toJSON} = render(<Input label="Email" leftIcon={faEnvelope} />);
    expect(toJSON()).toMatchSnapshot('Input_default');
  });

  it('Should call onChangeText while typing', async () => {
    const onChangeTextProp = jest.fn();
    const {getByTestId} = render(
      <Input label="Email" onChangeText={onChangeTextProp} />,
    );

    const value = 'fake_email@biruleibe.com';

    await act(async () => {
      const inputText = getByTestId(TEST_ID_INPUT);
      fireEvent.changeText(inputText, value);
      expect(onChangeTextProp).toBeCalledWith(expect.stringContaining(value));
    });
  });
});
