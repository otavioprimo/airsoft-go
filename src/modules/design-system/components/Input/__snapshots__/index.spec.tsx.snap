// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Design System - Input Should match snapshot with leftIcon: Input_default 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
        "paddingBottom": 16,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderBottomColor": "rgba(230,230,230,.9)",
          "borderBottomWidth": 1,
          "justifyContent": "center",
          "paddingBottom": 8,
          "paddingLeft": 8,
          "paddingRight": 8,
          "paddingTop": 8,
        },
      ]
    }
    testID="right_icon"
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight={19}
      bbWidth={19}
      fill="#fff"
      focusable={false}
      height={19}
      meetOrSlice={0}
      minX={0}
      minY={0}
      secondaryFill={null}
      secondaryOpacity={0.4}
      size={19}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          Object {},
          Object {
            "flex": 0,
            "height": 19,
            "width": 19,
          },
        ]
      }
      vbHeight={512}
      vbWidth={512}
      width={19}
    >
      <RNSVGGroup
        fill={4294967295}
        propList={
          Array [
            "fill",
          ]
        }
      >
        <RNSVGPath
          d="M464 64H48C21.49 64 0 85.49 0 112v288c0 26.51 21.49 48 48 48h416c26.51 0 48-21.49 48-48V112c0-26.51-21.49-48-48-48zm0 48v40.805c-22.422 18.259-58.168 46.651-134.587 106.49-16.841 13.247-50.201 45.072-73.413 44.701-23.208.375-56.579-31.459-73.413-44.701C106.18 199.465 70.425 171.067 48 152.805V112h416zM48 400V214.398c22.914 18.251 55.409 43.862 104.938 82.646 21.857 17.205 60.134 55.186 103.062 54.955 42.717.231 80.509-37.199 103.053-54.947 49.528-38.783 82.032-64.401 104.947-82.653V400H48z"
          fill={4294967295}
          fillOpacity={1}
          propList={
            Array [
              "fill",
              "fillOpacity",
            ]
          }
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
  <View
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "height": 38,
        },
      ]
    }
  >
    <Text
      animatedStyle={
        Object {
          "value": Object {
            "fontSize": 14,
            "transform": Array [
              Object {
                "translateY": 10,
              },
              Object {
                "translateX": 0,
              },
            ],
          },
        }
      }
      collapsable={false}
      style={
        Object {
          "color": "#fff",
          "fontSize": 14,
          "position": "absolute",
          "transform": Array [
            Object {
              "translateY": 10,
            },
            Object {
              "translateX": 0,
            },
          ],
        }
      }
    >
      Email
    </Text>
    <TextInput
      allowFontScaling={true}
      label="Email"
      leftIcon={
        Object {
          "icon": Array [
            512,
            512,
            Array [],
            "f0e0",
            "M464 64H48C21.49 64 0 85.49 0 112v288c0 26.51 21.49 48 48 48h416c26.51 0 48-21.49 48-48V112c0-26.51-21.49-48-48-48zm0 48v40.805c-22.422 18.259-58.168 46.651-134.587 106.49-16.841 13.247-50.201 45.072-73.413 44.701-23.208.375-56.579-31.459-73.413-44.701C106.18 199.465 70.425 171.067 48 152.805V112h416zM48 400V214.398c22.914 18.251 55.409 43.862 104.938 82.646 21.857 17.205 60.134 55.186 103.062 54.955 42.717.231 80.509-37.199 103.053-54.947 49.528-38.783 82.032-64.401 104.947-82.653V400H48z",
          ],
          "iconName": "envelope",
          "prefix": "far",
        }
      }
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      placeholderTextColor="rgba(230,230,230,.9)"
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "borderBottomColor": "rgba(230,230,230,.9)",
            "borderBottomWidth": 1,
            "color": "#fff",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "fontSize": 16,
          },
        ]
      }
      testID="input"
      underlineColorAndroid="transparent"
    />
  </View>
</View>
`;

exports[`Design System - Input Should match snapshot: Input_default 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
        "paddingBottom": 16,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "height": 38,
        },
      ]
    }
  >
    <Text
      animatedStyle={
        Object {
          "value": Object {
            "fontSize": 14,
            "transform": Array [
              Object {
                "translateY": 10,
              },
              Object {
                "translateX": 0,
              },
            ],
          },
        }
      }
      collapsable={false}
      style={
        Object {
          "color": "#fff",
          "fontSize": 14,
          "position": "absolute",
          "transform": Array [
            Object {
              "translateY": 10,
            },
            Object {
              "translateX": 0,
            },
          ],
        }
      }
    >
      Email
    </Text>
    <TextInput
      allowFontScaling={true}
      label="Email"
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      placeholderTextColor="rgba(230,230,230,.9)"
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "borderBottomColor": "rgba(230,230,230,.9)",
            "borderBottomWidth": 1,
            "color": "#fff",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "fontSize": 16,
          },
        ]
      }
      testID="input"
      underlineColorAndroid="transparent"
    />
  </View>
</View>
`;
